langchain/src/vectorstores/mongodb_atlas.ts

```diff
@@ -109,6 +110,84 @@ export class MongoDBAtlasVectorSearch extends VectorStore {
     return results.map((result) => result[0]);
   }
 
+  /**
+   * Return documents selected using the maximal marginal relevance.
+   * Maximal marginal relevance optimizes for similarity to the query AND diversity
+   * among selected documents.
+   *
+   * @param {string} query - Text to look up documents similar to.
+   * @param {number} k - Number of documents to return.
+   * @param {number} [fetchK=20]- Number of documents to fetch before passing to the MMR algorithm.
+   * @param {number} [lambda=0.5] - Number between 0 and 1 that determines the degree of diversity among the results,
+   *                 where 0 corresponds to maximum diversity and 1 to minimum diversity.
+   * @param {MongoDBDocument} preFilter - Optional Atlas Search operator to prefilter on document fields.
+   * @param {MongoDBDocument[]} postFilterPipeline - Optional pipeline of MongoDB aggregation stages following the knnBeta search.
+   *
+   * @returns {Promise<Document[]>} - List of documents selected by maximal marginal relevance.
+   */
```

Comment:
I know it's still in draft but would it make sense to have a switch in the constructor that uses this method by default when similaritySearch is called?

Reply:
Thank you for the review, @jacoblee93! I have tried to add the "useMMR" flag to the constructor args, but it didn't look very clean, as lambda and k mmr params also need to be specified. After discussing it with our team, we decided to create separate methods for mmr and similarity search.

It may be good to add maxMarginalRelevanceSearch method to the base VectorStore, the same way as in python implementation.

There is some code duplication to execute queries on the database, we will work to extract the common code to a private method as in https://github.com/langchain-ai/langchain/blob/1f40d3e0944604581f76a48fdfc961ac3017c0f2/libs/langchain/langchain/vectorstores/mongodb_atlas.py#L148

Reply:
Ah I see. Makes sense then! Let me know when this is ready for review.

Reply:
Made changes for MongoDBAtlasVectorSearch to use abstract similaritySearchVectorWithScore for MMR
Added optional maxMarginalRelevanceSearch method to base VectorStore
Think it's ready for review again @jacoblee93! 😀